### PR TITLE
raftstore: fix wrong last index

### LINKED ISSUE #1014: panic at 'slice out of range'
tikv log as follow:
```
2016-09-01 10:51:33,484 snap.rs:188 - INFO  - saving snapshot to /home/pingcap/work/poc/jinpeng/data2/snap/rev_188_6_114237.snap
2016-09-01 10:51:33,484 raft.rs:1265 - INFO  - [region 188] 190 [commit: 0, lastindex: 0, lastterm: 0] starts to restore snapshot [index: 114237, term: 6]
2016-09-01 10:51:33,484 raft.rs:1278 - INFO  - [region 188] 190 restored progress of 189 [Progress { matched: 0, next_idx: 1, state: Probe, paused: false, pending_snapshot: 0, recent_active: false, ins: Inflights { start: 0, count: 0, buffer: [] } }]
2016-09-01 10:51:33,484 raft.rs:1278 - INFO  - [region 188] 190 restored progress of 190 [Progress { matched: 0, next_idx: 1, state: Probe, paused: false, pending_snapshot: 0, recent_active: false, ins: Inflights { start: 0, count: 0, buffer: [] } }]
2016-09-01 10:51:33,484 raft.rs:1278 - INFO  - [region 188] 190 restored progress of 191 [Progress { matched: 0, next_idx: 1, state: Probe, paused: false, pending_snapshot: 0, recent_active: false, ins: Inflights { start: 0, count: 0, buffer: [] } }]
2016-09-01 10:51:33,484 raft_log.rs:385 - INFO  - log [committed=0, applied=0, unstable.offset=1, unstable.entries.len()=0] starts to restore snapshot [index: 114237, term: 6]
2016-09-01 10:51:33,484 raft.rs:1226 - INFO  - [region 188] 190 [commit: 114237] restored snapshot [index: 114237, term: 6]
2016-09-01 10:51:33,484 peer_storage.rs:378 - INFO  - [region 188] 190 begin to apply snapshot
2016-09-01 10:51:33,484 peer_storage.rs:407 - INFO  - [region 188] 190 apply snapshot for region id: 188 start_key: "t\200\000\000\000\000\000\000\377V_r\200\000\000\000\000\377\213+\365\000\000\000\000\000\372" end_key: "" region_epoch {conf_ver: 3 version: 33} peers {id: 189 store_id: 1} peers {id: 190 store_id: 4} peers {id: 191 store_id: 5} with state applied_index: 114237 truncated_state {index: 114237 term: 6} ok
2016-09-01 10:51:33,484 store.rs:736 - INFO  - [region 188] snapshot for region id: 188 start_key: "t\200\000\000\000\000\000\000\377V_r\200\000\000\000\000\377\213+\365\000\000\000\000\000\372" end_key: "" region_epoch {conf_ver: 3 version: 33} peers {id: 189 store_id: 1} peers {id: 190 store_id: 4} peers {id: 191 store_id: 5} is applied
2016-09-01 10:51:35,471 peer.rs:434 - WARN  - [region 152] 154 handle ready, entries 4, committed entries 1, messages 5, snapshot false, hard state changed true [takes Duration { secs: 1, nanos: 105315091 }]
2016-09-01 10:51:35,471 store.rs:588 - WARN  - on 1 regions raft ready [takes Duration { secs: 1, nanos: 105386342 }]
2016-09-01 10:51:35,784 snap.rs:186 - INFO  - apply new data takes Duration { secs: 1, nanos: 903820903 }
2016-09-01 10:51:35,795 mod.rs:51 - WARN  - handle task Snap apply for 180 [takes Duration { secs: 2, nanos: 453035375 }]
2016-09-01 10:51:35,795 snap.rs:122 - INFO  - begin apply snap data for 188
2016-09-01 10:51:35,953 snap.rs:188 - INFO  - saving snapshot to /home/pingcap/work/poc/jinpeng/data2/snap/rev_188_6_132154.snap
2016-09-01 10:51:35,953 raft.rs:1265 - INFO  - [region 188] 190 [commit: 114237, lastindex: 114237, lastterm: 6] starts to restore snapshot [index: 132154, term: 6]
2016-09-01 10:51:35,953 raft.rs:1278 - INFO  - [region 188] 190 restored progress of 189 [Progress { matched: 0, next_idx: 114238, state: Probe, paused: false, pending_snapshot: 0, recent_active: false, ins: Inflights { start: 0, count: 0, buffer: [] } }]
2016-09-01 10:51:35,953 raft.rs:1278 - INFO  - [region 188] 190 restored progress of 190 [Progress { matched: 114237, next_idx: 114238, state: Probe, paused: false, pending_snapshot: 0, recent_active: false, ins: Inflights { start: 0, count: 0, buffer: [] } }]
2016-09-01 10:51:35,953 raft.rs:1278 - INFO  - [region 188] 190 restored progress of 191 [Progress { matched: 0, next_idx: 114238, state: Probe, paused: false, pending_snapshot: 0, recent_active: false, ins: Inflights { start: 0, count: 0, buffer: [] } }]
2016-09-01 10:51:35,953 raft_log.rs:385 - INFO  - log [committed=114237, applied=114237, unstable.offset=114238, unstable.entries.len()=0] starts to restore snapshot [index: 132154, term: 6]
2016-09-01 10:51:35,953 raft.rs:1226 - INFO  - [region 188] 190 [commit: 132154] restored snapshot [index: 132154, term: 6]
thread 'raftstore-4' panicked at 'slice[132155,136093] out of bound[132155,132154]', src/raft/raft_log.rs:320
stack backtrace:
   1:     0x7f51cca2c02f - std::sys::backtrace::tracing::imp::write::h29f5fdb9fc0a7395
   2:     0x7f51cca3318b - std::panicking::default_hook::_{{closure}}::h2cc84f0378700526
   3:     0x7f51cca3159a - std::panicking::default_hook::hbbe7fa36a995aca0
   4:     0x7f51cc73a870 - tikv::util::panic_hook::set_exit_hook::_{{closure}}::h16f63522ff4f3661
                        at /home/pingcap/go/src/github.com/pingcap/tikv2/tikv/src/util/panic_hook.rs:68
   5:     0x7f51cca31bda - std::panicking::rust_panic_with_hook::h105c3d42fcd2fb5e
   6:     0x7f51cca31a41 - std::panicking::begin_panic::hbf62ea4a5ff3f9de
   7:     0x7f51cca3196a - std::panicking::begin_panic_fmt::h20f5943904e5791d
   8:     0x7f51cc51d303 - _<tikv..raft..raft_log..RaftLog<T>>::slice::hdad97e47cd6fba85
                        at /home/pingcap/go/src/github.com/pingcap/tikv2/tikv/<std macros>:8
                        at /home/pingcap/go/src/github.com/pingcap/tikv2/tikv/src/raft/raft_log.rs:339
   9:     0x7f51cc57badc - _<tikv..raftstore..store..store..Store<T, C>>::on_raft_ready::h32f5112b543f8d85
                        at /home/pingcap/go/src/github.com/pingcap/tikv2/tikv/src/raft/raft_log.rs:292
                        at /home/pingcap/go/src/github.com/pingcap/tikv2/tikv/src/raft/raw_node.rs:112
                        at /home/pingcap/go/src/github.com/pingcap/tikv2/tikv/src/raft/raw_node.rs:282
                        at /home/pingcap/go/src/github.com/pingcap/tikv2/tikv/src/raftstore/store/peer.rs:408
                        at /home/pingcap/go/src/github.com/pingcap/tikv2/tikv/src/raftstore/store/store.rs:568
  10:     0x7f51cc58e199 - _<tikv..raftstore..store..store..Store<T, C>>::run::h5721048789679458
                        at /home/pingcap/go/src/github.com/pingcap/tikv2/tikv/src/raftstore/store/store.rs:1336
                        at /home/pingcap/.cargo/git/checkouts/mio-e79f285e2377d154/master/src/event_loop.rs:315
                        at /home/pingcap/.cargo/git/checkouts/mio-e79f285e2377d154/master/src/event_loop.rs:261
                        at /home/pingcap/go/src/github.com/pingcap/tikv2/tikv/src/raftstore/store/store.rs:242
  11:     0x7f51cc5deeb4 - _<tikv..server..node..Node<C>>::start_store::_{{closure}}::h3c432f9f1415293b
  12:     0x7f51cc5dc07a - std::panicking::try::_{{closure}}::_{{closure}}::hcf50daaf458fd93a
  13:     0x7f51cca3a99b - __rust_try
  14:     0x7f51cca3a87e - __rust_maybe_catch_panic
  15:     0x7f51cc5db2ae - std::thread::Builder::spawn::_{{closure}}::h23348431e6569cfb
  16:     0x7f51cc568aed - _<F as alloc..boxed..FnBox<A>>::call_box::h638735dc521642c1
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/liballoc/boxed.rs:587
  17:     0x7f51cca2fc64 - std::sys::thread::Thread::new::thread_start::h8f3bd45211e9f5ea
  18:     0x7f51cb151183 - start_thread
  19:     0x7f51cb96b37c - clone
  20:                0x0 - <unknown>
```

### PR DESCRIPTION
**Pull request by BusyJay**
When applying two snapshot async at the same time, raft state machine may compute the wrong last index. So we need to abort applying old snapshot first.

close #1014

### CODE REVIEW

File: src/raftstore/store/msg.rs [hunk 1]
```diff
@@ -60,7 +60,11 @@ pub enum Msg {
 
     // For snapshot stats.
     SnapshotStats,
-    SnapApplyRes { region_id: u64, is_success: bool },
+    SnapApplyRes {
+        region_id: u64,
+        is_success: bool,
+        is_abort: bool,
```
Comment (siddontang) [inline src/raftstore/store/msg.rs]: /abort/aborted/

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -408,6 +408,13 @@ impl Peer {
         let mut ready = self.raft_group.ready();
         let is_applying = self.get_store().is_applying_snap();
         if is_applying {
+            if !raft::is_empty_snap(&ready.snapshot) {
+                if !self.get_store().is_canceling_snap() {
+                    warn!("receiving a new snap when applying the old one, try to abort.");
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: can we output last snapshot's meta and the new one's?

File: src/raftstore/store/worker/snap.rs [hunk 3]
```diff
@@ -70,6 +78,58 @@ impl MsgSender for SendCh<Msg> {
...
+    Ok(())
+}
+
+fn delete_all_in_range(db: &DB,
+                       start_key: &[u8],
+                       end_key: &[u8],
+                       abort: &AtomicBool)
+                       -> Result<(), Error> {
+    for cf in db.cf_names() {
+        try!(check_abort(&abort));
```
Comment (siddontang) [inline src/raftstore/store/worker/snap.rs]: need to check abort multi times?

Comment (BusyJay) [inline src/raftstore/store/worker/snap.rs]: Yes. To abort as fast as we can.

File: src/raftstore/store/peer_storage.rs [hunk 4]
```diff
@@ -45,15 +45,32 @@ const MAX_SNAP_TRY_CNT: usize = 5;
...
 
-#[derive(PartialEq, Debug)]
+#[derive(Debug)]
 pub enum SnapState {
     Relax,
     Generating,
     Snap(Snapshot),
-    Applying,
+    Applying(Arc<AtomicBool>),
+    ApplyAbort,
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: ApplyAborted

File: src/raftstore/store/store.rs [hunk 5]
```diff
@@ -543,7 +548,9 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         }
         for region in &self.pending_regions {
             if enc_start_key(region) < enc_end_key(&snap_region) &&
-               enc_end_key(region) > enc_start_key(&snap_region) {
+               enc_end_key(region) > enc_start_key(&snap_region) &&
+               // Same region can overlap, we will apply the latest version of snapshot.
+               region.get_id() != snap_region.get_id() {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: no need to handle id equality?

File: src/raftstore/store/worker/snap.rs [hunk 6]
```diff
@@ -187,33 +251,46 @@ impl<T: MsgSender> Runner<T> {
...
         let ts = Instant::now();
-        let mut is_success = true;
-        if let Err(e) = self.apply_snap(region_id) {
-            is_success = false;
-            error!("failed to apply snap: {:?}!!!", e);
-        }
+        let (is_success, is_aborted) = match self.apply_snap(region_id, abort) {
+            Ok(()) => (true, false),
+            Err(Error::Abort) => {
+                warn!("applying snapshot for region {} is abort.", region_id);
```
Comment (siddontang) [inline src/raftstore/store/worker/snap.rs]: is aborted

File: src/raftstore/store/store.rs (hunk 5 above)
Comment (BusyJay) [inline src/raftstore/store/store.rs]: No, raft can handle it automatically.

File: src/raftstore/store/worker/snap.rs [hunk 7]
```diff
@@ -147,13 +208,15 @@ impl<T: MsgSender> Runner<T> {
...
         }
+        try!(check_abort(&abort));
         box_try!(snap_file.validate());
         let mut reader = box_try!(File::open(snap_file.path()));
 
         let timer = Instant::now();
         // Write the snapshot into the region.
         loop {
             // TODO: avoid too many allocation
+            try!(check_abort(&abort));
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/snap.rs]: Is it more reasonable check_abort before every db.write(wb)? Just a suggestion.

Comment (BusyJay) [inline src/raftstore/store/worker/snap.rs]: Its goal is to abort as soon as possible rather than to avoid possible time consuming operation.

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @hhkbp2

File: src/raftstore/store/peer_storage.rs [hunk 8]
```diff
@@ -490,9 +507,37 @@ impl PeerStorage {
...
+
+    #[inline]
+    pub fn is_canceling_snap(&self) -> bool {
+        match *self.snap_state.borrow() {
+            SnapState::Applying(ref abort) => abort.load(Ordering::Relaxed),
+            _ => false,
+        }
+    }
+
+    pub fn cancle_applying_snap(&mut self) {
```
Comment (hhkbp2) [inline src/raftstore/store/peer_storage.rs]: cancle -> cancel

File: src/raftstore/store/worker/snap.rs (hunk 7 above)
Comment (hhkbp2) [inline src/raftstore/store/worker/snap.rs]: Any specified reason why we need to choose fast abort over a clean code?

Comment (BusyJay) [inline src/raftstore/store/worker/snap.rs]: Otherwise the memory may likely be run out.

File: src/raftstore/store/peer.rs [hunk 9]
```diff
@@ -408,6 +408,14 @@ impl Peer {
         let mut ready = self.raft_group.ready();
         let is_applying = self.get_store().is_applying_snap();
         if is_applying {
+            if !raft::is_empty_snap(&ready.snapshot) {
+                if !self.get_store().is_canceling_snap() {
+                    warn!("receiving a new snap {:?} when applying the old one, try to abort.",
+                          ready.snapshot);
+                    self.mut_store().cancle_applying_snap();
```
Comment (hhkbp2) [inline src/raftstore/store/peer.rs]: A hypothetical case:

Applying snapshot is slow for follower in a server heavy load. And the leader re-sends the same snapshot in a fast speed. What will happen then? Is there any chance(reasonable?) the follower could actually have enough time to apply the snapshot?

Comment (BusyJay) [inline src/raftstore/store/peer.rs]: No, the same snapshot won't be sent again once it's applied in the state machine.

**Comment by hhkbp2** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**